app/Http/Controllers/Product/ExtendedBaseProductController.php

```diff
@@ -149,17 +150,35 @@ public function adminDownload($id, $invoice = '', $api = false)
                 header('Content-Length: '.filesize($release));
                 readfile($release);
                 // ob_end_clean();
+                }
+            } else {
+               throw new \Exception(\Lang::get('message.no_permission_for_action'));
             }
-        } catch (\Exception $e) {
-            if ($api) {
-                return response()->json(['error'=>$e->getMessage()]);
-            }
+           } catch (\Exception $e) {
             Bugsnag::notifyException($e);
-
             return redirect()->back()->with('fails', $e->getMessage());
         }
     }
 
+    /**
+     * Checks whether order exists or not for a product and invoice
+     *
+     * @date   2020-04-13T14:53:04+0530
+     *
+     * @param  int $id                  Product id
+     * @param  int $invoice             Invoice Number
+     * @param  boolean $allowDownload     
+     * @return boolean
+     */
+    private function downloadValidation(bool $allowDownload,$id,$invoice) 
+    {
+        if(\Auth::user()->role =='user') {
+        $invoice = Invoice::where('number', $invoice)->first();//If invoice number sent as parameter exists
+        $allowDownload = $invoice ? $invoice->order()->value('product') == $id  : false; //If the order for the product sent in the parameter exists
```

Comment:
do we even need to check for product id? Can't we send download response based on order_id directly?

Reply:
If the product in the order table is different with the one sent in request($id), then only it will return true

Reply:
but do we even need that?

Reply:
yes we need to verify the id sent in the request with the product in orders table brfore allowing to download